### ISSUE TITLE
The usage of the "forced_decoder_ids" parameter

### ISSUE THREAD

**Issue by Alsace08** [huggingface/transformers#34808]
### Feature request

How to use the "forced_decoder_ids" parameter for decoder-only models? This parameter seems to be deprecated in the latest version.
![微信图片_20241119222811](https://github.com/user-attachments/assets/a5f2d44f-d98f-4959-85ae-e64489e3b9df)

### Motivation

This is an important function.

### Your contribution

This is an important function.

**Comment by zucchini-nlp** [comment]
@Alsace08 hey! If you are trying to use it with Whisper models, then it is recommended to indicate the task/language/decoder_input_ids using the keywords  task/language/prompt_ids when you are calling `generate()`. For ex:

`model.generate(input_features, task=task, language=language, prompt_ids=prompt_ids)`

**Comment by Alsace08** [comment]
> @Alsace08 hey! If you are trying to use it with Whisper models, then it is recommended to indicate the task/language/decoder_input_ids using the keywords task/language/prompt_ids when you are calling `generate()`. For ex:
> 
> `model.generate(input_features, task=task, language=language, prompt_ids=prompt_ids)`

I'm trying the Llama model, any solutions?

**Comment by zucchini-nlp** [comment]
@Alsace08 I see. Since we had the `forced_decoder_ids` added and supported only for Whisper, it was deprecated in favor of other kwargs. 

If you want to force certain tokens in Llama you can try [constrained decoding](https://huggingface.co/blog/constrained-beam-search) or `PrefixConstrainedLogitsProcessor` depending in your use case. Otherwise you can write your own custom logits processors and pass them down to the `generate()`. Would that work for you?

**Comment by qgallouedec** [comment]
Hey @Alsace08, sorry for the slow follow-up here, we're catching up on the backlog.

`forced_decoder_ids` is a Whisper-only mechanism see
https://github.com/huggingface/transformers/blob/c7ad72879438c7d83955394312cea9f437e9f43b/src/transformers/models/whisper/generation_whisper.py#L1491-L1499
that was deprecated in favor of the `task`/`language` kwargs, so it was never wired up for decoder-only models like Llama and we don't plan to extend it there.

For forcing specific tokens during generation on a causal LM, use constrained/prefix-constrained decoding instead, see `PrefixConstrainedLogitsProcessor` (https://github.com/huggingface/transformers/blob/c7ad72879438c7d83955394312cea9f437e9f43b/src/transformers/generation/logits_process.py#L1484-L1495
or the constrained beam search guide (https://huggingface.co/blog/constrained-beam-search), or write a custom `LogitsProcessor` and pass it via `logits_processor=` to `generate()`.
